### PR TITLE
Feature/service discovery

### PR DESCRIPTION
**Pull request by idohuber**
## Overview
This PR adds #
service scanner for applications authentication security

**[Signed Commits](../CONTRIBUTING.md#sign-off-per-commit)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: servicehandler/servicediscovery.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+type ServiceAuthentication struct {
+	kind       string
+	apiVersion string
+	metadata   struct {
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:56]: consider extracting

File: servicehandler/servicediscovery.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+			// get a list of all  services in the cluster
+			services, err := regularClient.Services("").List(context.TODO(), serviceListOptions)
+			if err != nil {
+				logger.L().Ctx(ctx).Error(err.Error())
+				return
+			}
+
+			scansWg := sync.WaitGroup{}
+			antsPool, _ := ants.NewPool(20)
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:186]: check error

File: servicehandler/servicediscovery.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+			for _, service := range services.Items {
+				sra := ServiceAuthentication{}
+				sra.kind = kind
+				sra.apiVersion = apiVersion
+				sra.metadata.name = service.Name
+				sra.metadata.namespace = service.Namespace
+				sra.spec.clusterIP = service.Spec.ClusterIP
+				sra.initialPorts(service.Spec.Ports)
+
+				currentServiceList = append(currentServiceList, [2]string{service.Name, service.Namespace})
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:197]: v = []string{service.Name, service.Namespace}

File: servicehandler/servicediscovery.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+				continue
+			}
+			logger.L().Ctx(ctx).Info("Authentication Service " + service.GetName() + " in namespace " + service.GetNamespace() + " deleted")
+		}
+	}
+	return err
+}
+
+func DiscoveryServiceHandler(ctx context.Context, kubeClient *k8sinterface.KubernetesApi, timeout time.Duration) {
+	{
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:170]: 2 {

File: servicehandler/servicediscovery.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	// get all services from the current cycle and compare them with the current CRDs
+
+	authServices, err := client.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, service := range authServices.Items {
+		if !csl.contains(service.GetName(), service.GetNamespace()) {
+			err := client.Namespace(service.GetNamespace()).Delete(context.TODO(), service.GetName(), metav1.DeleteOptions{})
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:158]: you have the context from func argument

File: servicehandler/servicediscovery.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	FieldSelector: "metadata.namespace!=kube-system",
+}
+
+var Schema = schema.GroupVersionResource{
+	Group:    group,
+	Version:  version,
+	Resource: resource,
+}
+
+type currentServiceList [][2]string
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:42]: replace with struct

File: servicehandler/servicediscovery.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+	serviceObj, structuredErr := sra.Unstructured()
+	if structuredErr != nil {
+		logger.L().Ctx(ctx).Error(structuredErr.Error())
+		return
+	}
+
+	_, deleteErr := client.Namespace(sra.metadata.namespace).Apply(context.TODO(), sra.metadata.name, serviceObj, metav1.ApplyOptions{FieldManager: FieldManager})
+	if deleteErr != nil {
+		logger.L().Ctx(ctx).Error(deleteErr.Error())
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:121]: no need to return the error?

File: servicehandler/servicediscovery.go (hunk 3 above)
Comment (RinaO1234) [reply servicehandler/servicediscovery.go:197]: use struct

File: servicehandler/servicediscovery.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func DiscoveryServiceHandler(ctx context.Context, kubeClient *k8sinterface.KubernetesApi, timeout time.Duration) {
+	{
+		dynamicClient := kubeClient.DynamicClient.Resource(Schema)
+		regularClient := kubeClient.KubernetesClient.CoreV1()
+
+		for {
+			logger.L().Ctx(ctx).Info("starting a new service discovery handling")
+			currentServiceList := make(currentServiceList, 0)
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:176]: if capacity is already known, allocate it here to avoid redundant allocatations with append

File: servicehandler/servicediscovery.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+	FieldManager = "kubescape|operator|serviceDiscoveryHandler"
+)
+
+var protocolFilter = []string{"UDP"}
+
+var serviceListOptions = metav1.ListOptions{
+	FieldSelector: "metadata.namespace!=kube-system",
+}
+
+var Schema = schema.GroupVersionResource{
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:36]: is Schema public use? if no, should be private like the above vars

File: servicehandler/servicediscovery.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+		return
+	}
+
+	_, deleteErr := client.Namespace(sra.metadata.namespace).Apply(context.TODO(), sra.metadata.name, serviceObj, metav1.ApplyOptions{FieldManager: FieldManager})
+	if deleteErr != nil {
+		logger.L().Ctx(ctx).Error(deleteErr.Error())
+	}
+}
+
+func (port *Port) Scan(ctx context.Context, ip string) {
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:125]: consider dividing these types to other files

File: servicehandler/servicediscovery.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	result, err := cmd.ScanTargets(ctx, ip, port.port)
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:126]: error isn't checked

File: servicehandler/servicediscovery.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+	port.applicationLayer = result.ApplicationLayer
+	port.presentationLayer = result.PresentationLayer
+	port.sessionLayer = result.SessionLayer
+	port.authenticated = result.IsAuthenticated
+
+	if result.ApplicationLayer == "" {
+		// if we can't get the application layer, then we change to Unknown
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:133]: consider adding debug message

**Review by matthyx — CHANGES_REQUESTED** [review]

File: config/config.go [hunk 13]
```diff
@@ -47,6 +47,11 @@ type Components struct {
 	Storage            Component `mapstructure:"storage"`
 }
 
+type ServiceScanConfig struct {
```
[the remark this answers is not in the document]
Comment (idohuber) [reply config/config.go:50]: i didnt find where to add it

File: servicehandler/servicediscovery_test.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+// 	logger.L().Ctx(ctx).Info("finished service discovery cycle")
+// }
+
+//TODO: Add more tests
+
+//TODO use fake client
+
+//TODO add tests to strict functionallity
+
+//TODO use kwok for perfomnace by using scale
```
[the remark this answers is not in the document]
Comment (idohuber) [reply servicehandler/servicediscovery_test.go:141]: ok got it

File: servicehandler/servicediscovery_test.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+				sessionLayer:      "test",
+				presentationLayer: "test",
+				applicationLayer:  "test",
+				authenticated:     true,
+			},
+		},
+	},
+}
+
+func TestUnstructured(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (idohuber) [reply servicehandler/servicediscovery_test.go:47]: why not?
because it more for error handling than test?

File: config/config.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (idohuber) [reply config/config.go:50]: because the tests are from the file , and i get the config from the config map

File: servicehandler/servicediscovery_test.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+func TestContains(t *testing.T) {
+	services := currentServiceList{
+		{"test", "test"},
+	}
+	if !services.contains("test", "test") {
+		t.Errorf("contains() returned false")
+	}
+}
+
+func TestInitialPorts(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (idohuber) [reply servicehandler/servicediscovery_test.go:61]: i dont think its super important ,
but for checking functions expectations?

File: servicehandler/servicediscovery_test.go [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+			inObjects := append(tc.services, tc.delete...)
+			regClient := kubernetesFake.NewSimpleClientset(inObjects...)
+
+			var crds *unstructured.UnstructuredList
+			for i := 0; i < 2; i++ {
+				services, _ := serviceExtractor(ctx, regClient)
+				for _, service := range services {
+					obj, _ := service.unstructured()
+					_, err := dynamicClient.Resource(ServiceScanSchema).Namespace(service.metadata.namespace).Create(ctx, obj, metav1.CreateOptions{})
+					require.NoError(t, err)
```
Comment (matthyx) [inline servicehandler/servicediscovery_test.go:306]: make sure we can run with these

File: servicehandler/servicediscovery_test.go [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+				}
+				err := discoveryService(context.Background(), regClient, dynamicClient)
+				assert.NoError(t, err)
+
+				crds, _ = dynamicClient.Resource(ServiceScanSchema).List(ctx, metav1.ListOptions{})
+
+				if tc.delete != nil {
+					for _, delService := range tc.delete {
+						err := regClient.CoreV1().Services(delService.(*v1.Service).Namespace).Delete(ctx, delService.(*v1.Service).Name, metav1.DeleteOptions{})
+						require.NoError(t, err)
```
Comment (matthyx) [inline servicehandler/servicediscovery_test.go:316]: and here too

**Review by matthyx — CHANGES_REQUESTED** [review]
I've improved the tests, see if you can make them run without commenting the NoErrors

**Review by matthyx — APPROVED** [review]

File: servicehandler/portsscan.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+	"context"
+
+	logger "github.com/kubescape/go-logger"
+	"github.com/kubescape/kubescape-network-scanner/cmd"
+	v1 "k8s.io/api/core/v1"
+)
+
+type Port struct {
+	port              int
+	protocol          string
```
Comment (RinaO1234) [inline servicehandler/portsscan.go:13]: would be nicer to have it as defined type with set of values, or iota

File: servicehandler/portsscan.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	sessionLayer      string
+	presentationLayer string
+	applicationLayer  string
+	authenticated     bool
+}
+
+func (port *Port) scan(ctx context.Context, ip string) {
+	result, err := cmd.ScanTargets(ctx, ip, port.port)
+	port.applicationLayer = result.ApplicationLayer
```
Comment (RinaO1234) [inline servicehandler/portsscan.go:22]: error is skipped

File: servicehandler/portsscan.go [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+	port              int
+	protocol          string
+	sessionLayer      string
+	presentationLayer string
+	applicationLayer  string
+	authenticated     bool
+}
+
+func (port *Port) scan(ctx context.Context, ip string) {
+	result, err := cmd.ScanTargets(ctx, ip, port.port)
```
Comment (RinaO1234) [inline servicehandler/portsscan.go:21]: error is unchecked

File: servicehandler/portsscan.go [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+	port.applicationLayer = result.ApplicationLayer
+	port.presentationLayer = result.PresentationLayer
+	port.sessionLayer = result.SessionLayer
+	port.authenticated = result.IsAuthenticated
+
+	if result.ApplicationLayer == "" {
+		// if we can't get the application layer, then we change to Unknown
```
Comment (RinaO1234) [inline servicehandler/portsscan.go:28]: debug message would be nice

File: servicehandler/portsscan.go [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+		port.authenticated = true
+	}
+
+	if err != nil {
+		//if we have an error, we log it and set all layers to Unknown
+		logger.L().Ctx(ctx).Error(err.Error())
+		port.applicationLayer = "failed_to_scan"
```
Comment (RinaO1234) [inline servicehandler/portsscan.go:35]: better define as a status where each status can have also string message to indicate what happened

File: servicehandler/servicediscovery.go [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+	for _, service := range authServices.Items {
+		crdMetadata := metadata{
+			name:      service.GetName(),
+			namespace: service.GetNamespace(),
+		}
+
+		if !slices.Contains(currentServicesMetadata, crdMetadata) {
+			err := client.Namespace(service.GetNamespace()).Delete(ctx, service.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				logger.L().Ctx(ctx).Error(err.Error())
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:61]: if there's an error, continue to next item  or return error (decide based on the logic)

File: servicehandler/servicediscovery_test.go [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	dynamicFake "k8s.io/client-go/dynamic/fake"
+	kubernetesFake "k8s.io/client-go/kubernetes/fake"
+)
+
+var TestAuthentications = serviceAuthentication{
```
Comment (RinaO1234) [inline servicehandler/servicediscovery_test.go:19]: ensure adding bad scenarios tests such as empty inputs or nil feilds

File: servicehandler/portsscan.go (hunk 22 above)
Comment (idohuber) [reply servicehandler/portsscan.go:28]: dont agree there is no enough context for the user , 
and in the scanner POV if i cant discover the app its mean it is safe from me
th elogic for this will be in the control

File: servicehandler/portsscan.go (hunk 23 above)
Comment (idohuber) [reply servicehandler/portsscan.go:35]: can you explain more what do you mean please?

File: servicehandler/portsscan.go (hunk 19 above)
Comment (idohuber) [reply servicehandler/portsscan.go:13]: can you explain further please?

File: servicehandler/servicediscovery.go [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+	return currentServiceList, metadataList
+
+}
+
+func discoveryService(ctx context.Context, regularClient kubernetes.Interface, dynamicClient dynamic.Interface) error {
+	serviceList, metadataList := serviceExtractor(ctx, regularClient)
+
+	scanWg := sync.WaitGroup{}
+	p, err := ants.NewPoolWithFunc(workerNum, func(i interface{}) {
+		sra := i.(serviceAuthentication)
```
Comment (RinaO1234) [inline servicehandler/servicediscovery.go:164]: add sra , ok because otherwise it may panic
